### PR TITLE
Public dependency analyzer

### PR DESCRIPTION
**Pull request by evildour**
This analyzer is disabled by default and needs to be opted into. I added a ruleset file to do just that for the gRPC projects. I'm having trouble building everything locally due to #1455 (not sure if that's just a problem on my machine or something wrong with that PR), but I'm creating this PR to see what the build machines do.

### CODE REVIEW

**Comment by benwulfe** [comment]
Looks like you may have resolved a merge conflict incorrectly  I will
highlight on the PR.

File: apis/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1.csproj [hunk 1]
```diff
@@ -21,12 +21,11 @@
     <PackageProjectUrl>https://github.com/GoogleCloudPlatform/google-cloud-dotnet</PackageProjectUrl>
     <RepositoryType>git</RepositoryType>
     <RepositoryUrl>https://github.com/GoogleCloudPlatform/google-cloud-dotnet</RepositoryUrl>
+    <CodeAnalysisRuleSet>..\..\..\grpc.ruleset</CodeAnalysisRuleSet>
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="ConfigureAwaitChecker.Analyzer" Version="1.0.0-beta4" PrivateAssets="All" />
     <PackageReference Include="Google.Api.Gax.Grpc" Version="2.1.0-beta02" />
-    <PackageReference Include="Google.Cloud.Iam.V1" Version="1.0.0-beta12" />
```
Comment (benwulfe) [inline apis/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1.csproj]: Looks like this was a merge conflict resolution issue.  this should have stayed around

File: apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1.csproj [hunk 2]
```diff
@@ -21,12 +21,11 @@
     <PackageProjectUrl>https://github.com/GoogleCloudPlatform/google-cloud-dotnet</PackageProjectUrl>
     <RepositoryType>git</RepositoryType>
     <RepositoryUrl>https://github.com/GoogleCloudPlatform/google-cloud-dotnet</RepositoryUrl>
+    <CodeAnalysisRuleSet>..\..\..\grpc.ruleset</CodeAnalysisRuleSet>
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="ConfigureAwaitChecker.Analyzer" Version="1.0.0-beta4" PrivateAssets="All" />
     <PackageReference Include="Google.Api.Gax.Grpc" Version="2.1.0-beta02" />
-    <PackageReference Include="Google.Cloud.Iam.V1" Version="1.0.0-beta12" />
```
Comment (benwulfe) [inline apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1.csproj]: Same as below -- this should not be deleted.

File: apis/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1.csproj (hunk 1 above)
Comment (evildour) [reply apis/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1/Google.Cloud.Spanner.Admin.Instance.V1.csproj]: I don't think it was a conflict. See my comment on #1455. Re-running the project generator automatically removed these.

File: tools/Google.Cloud.Tools.Analyzers/PublicDependencyForbiddenAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        private static TypeCheckResult CheckType(ITypeSymbol type, Dictionary<ITypeSymbol, TypeCheckResult> checkedTypes)
+        {
+            if (!checkedTypes.TryGetValue(type, out var result))
+            {
+                // Add a placeholder before we do anything else to prevent recursion issues.
+                checkedTypes.Add(type, TypeCheckResult.Allowed);
+
+                result = CheckTypeImpl(type);
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.Analyzers/PublicDependencyForbiddenAnalyzer.cs]: This is a really interesting use for local methods (do something with the return value in every case, even if there are multiple return statements). May put this into C# in Depth :)

Comment (evildour) [reply tools/Google.Cloud.Tools.Analyzers/PublicDependencyForbiddenAnalyzer.cs]: Thanks. I'm not sure if it's the cleanest way to do it, but I like how it let me memoize some functionality without exposing the raw functionality on the class and without having to introduce a whole bunch of extra if/elses.

**Pull request merged**